Support Metrics mode when creating/writing Iceberg tables

liqinrae opened the pull request:
Added support for all metrics modes in orc format as mentioned in #9791 

1. Default metrics mode is changed to truncate(16)
2. Allow for configuring default mode to NONE so that we can skip all metrics by default except the ones we want
3. Allows per column metrics mode configuration to support different modes: counts, none, truncate(length) and full

rdsr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -116,15 +124,22 @@ private static Metrics computeMetrics(Schema icebergSchema, ColumnMetadata<OrcTy
-                upperBoundsBuilder.put(icebergId, minMax.getMax());
-            });
+
+            if (metricsMode != MetricsModes.Counts.get()) {
+                toIcebergMinMax(orcColumnStats, icebergField.type(), metricsMode).ifPresent(minMax -> {
```
Do we not handle metrics mode `truncate`?

liqinrae replied:
@rdsr yes we do, we handle truncate in the `IcebergMinMax `

rdsr replied:
I see we added support for this below

jackye1995 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -431,7 +432,8 @@ public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, Con
                 getFileFormat(transaction.table()),
-                transaction.table().properties());
+                transaction.table().properties(),
+                Collections.emptyMap());
```
the whole table property map is already passed in as a part of the writable table handle, why do we need to construct the metrics config as a separated field?

liqinrae replied:
@jackye1995 yes you are right. There is no need to construct the metrics config since the table properties have already been passed to the writable table handle. I checked the code and we already have functions in place to retrieve metrics related properties so we can make this even simpler. Will add a fix to this, thanks!

jackye1995 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -128,6 +131,7 @@ public IcebergPageSink(
         this.jsonCodec = requireNonNull(jsonCodec, "jsonCodec is null");
         this.session = requireNonNull(session, "session is null");
         this.fileFormat = requireNonNull(fileFormat, "fileFormat is null");
+        this.metricsConfig = MetricsConfig.fromProperties(requireNonNull(storageProperties, "metricsConfig is null"));
```
error message does not match. 

Btw, I think it makes more sense to pass in the `MetricsConfig` to the constructor instead of `storageProperties`, because file format is also derived from `storageProperties` and we are passing it in separately.

liqinrae replied:
@jackye1995 Sorry you meant passing in the `MetricsConfig` to the `pagesink` or `IcebergWritableTableHandle` constructor? 

The file format is derived from the property map and then passed to the writable table handle as a separate component, which is how we passed in`MetricsConfig` initially.

jackye1995 replied:
to the page sink. But I don't have a strong opinion on this, there are many duplicated information passed across the codebase as of today. We can just fix the error message first.

liqinrae replied:
> @liqinrae can you please check the Ci build results?

@findepi Yes seems like there is new changes from other PRs. Will merge them and fix the CI build.

findepi commented:
@liqinrae can you please check the Ci build results?

jackye1995 reviewed (approved):
looks good to me

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergFileWriterFactory.java:
```diff
@@ -110,13 +111,14 @@ public IcebergFileWriter createFileWriter(
                 return createParquetWriter(outputPath, icebergSchema, jobConf, session, hdfsContext);
             case ORC:
-                return createOrcWriter(outputPath, icebergSchema, jobConf, session);
+                return createOrcWriter(metricsConfig, outputPath, icebergSchema, jobConf, session);
```
what about Parquet?

jackye1995 replied:
I can publish a followup for parquet.

findepi replied:
@liqinrae add `// TODO use metricsConfig` in the PARQUET branch

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -128,6 +131,7 @@ public IcebergPageSink(
         this.jsonCodec = requireNonNull(jsonCodec, "jsonCodec is null");
         this.session = requireNonNull(session, "session is null");
         this.fileFormat = requireNonNull(fileFormat, "fileFormat is null");
+        this.metricsConfig = MetricsConfig.fromProperties(requireNonNull(storageProperties, "storageProperties is null"));
```
https://github.com/apache/iceberg/pull/3638

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSinkProvider.java:
```diff
@@ -91,6 +91,7 @@ private ConnectorPageSink createPageSink(ConnectorSession session, IcebergWritab
                 jsonCodec,
                 session,
                 tableHandle.getFileFormat(),
+                tableHandle.getStorageProperties(),
```
how is a user expected to set storage properties?

liqinrae replied:
@findepi The whole table property map is passed to the writable file handle in `IcebergMetadata`. I am passing it to the `PageSink` constructor to derive the metrics config property.

findepi replied:
sure, but that doesn't answer my question -- how a Trino user can set this? how an Iceberg user would do that?

jackye1995 replied:
Users can set it in Spark.

findepi replied:
_how_ do they do that?

then, if this is useful, _how_ users will do this via Trino?

jackye1995 replied:
Using these table properties: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableProperties.java#L187-L189

And the compute engine or user calls the Iceberg API like:

```java
table.updateProperties()
.set("write.metadata.metrics.default", "full")
.commit()
```

I think just respecting the metrics configuration on Trino side is a good start in case the same config is set in Spark. I don't know if you would want users to set this on Trino side, because it seems like you want the file writing process to be more managed instead of users changing the write configs here and there. But if Trino is open for configuring this part, then we can use the same mechanism and introduce corresponding Iceberg table properties, maybe through 1 property for table default, 1 map property for column defaults.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -118,15 +126,22 @@ private static Metrics computeMetrics(Schema icebergSchema, ColumnMetadata<OrcTy
-                upperBoundsBuilder.put(icebergId, minMax.getMax());
-            });
+
+            if (metricsMode != MetricsModes.Counts.get()) {
```
`MetricsMode` is an interface, open for extensions. The fact that something isn't Counts doesn't mean it's Full.

Also, there is no javadoc indicating `.get()` returns same instance every time, so using `.equals()` would be more appropriate here.

this should be handled in a defensive manner:

```
if (metricsMode.equals(MetricsModes.None.get())) {
 // nothing to do
}
else if (metricsMode.equals(MetricsModes.Counts.get()) {
 .. collect counts 
}
else if (metricsMode.equals( full )) {
  ...
}
else {
    throw new UnsupportedOperationException("Unsupported metrics mode: " + metricsMode);
}
```

liqinrae replied:
@findepi The conditional statement here handles the case where the metricsMode is truncate(length) or full. Both of these two cases need to store min/max except that truncate need to apply truncation if needed. 

All of the metrics modes other than None need to collect counts. There would be duplication of code if we are to handle each mode separately. Can we add a `verify` function instead to make sure that the metric mode is supported if we want to handle it more defensively?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -261,10 +276,20 @@ private static int getIcebergId(OrcType orcColumn)
-            this.max = Conversions.toByteBuffer(type, max);
+            // Out of the two types which could be truncated, string or binary, ORC only supports string bounds.
+            // Therefore, truncation will be applied if needed only on string type.
+            if (metricsMode instanceof MetricsModes.Truncate && type.typeId() == org.apache.iceberg.types.Type.TypeID.STRING) {
```
binary (varbinary)  and fixed too 

https://github.com/apache/iceberg/blob/2e4684740b56f6d80328f50b7e7fef0863c262e7/parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java#L269

(we don't support creating tables with `fixed`, but i think we do support writes to such tables)

liqinrae replied:
@findepi I was referring to the `OrcMetrics` implementation here: https://github.com/apache/iceberg/blob/ddc5aff9167c7e367e2b4313b91e55451f09ef16/orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java#L281 
Seems like it only added truncation for string bound in the orc format.

findepi replied:
because they don't produce any stats for binary for ORC, see https://github.com/trinodb/trino/blob/88ef2caf7b5f374c7558e2f06df754c2d6b1acac/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcConnectorTest.java#L31 
right?

findepi replied:
still, not having an `if` for varbinary doesn't seem future proof, does it?

liqinrae replied:
@findepi yes I agree. I'll add varbinary here.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -261,10 +276,20 @@ private static int getIcebergId(OrcType orcColumn)
+                this.max = UnicodeUtil.truncateStringMax(Literal.of((CharSequence) max), truncateLength).toByteBuffer();
+            }
+            else {
+                this.min = Conversions.toByteBuffer(type, min);
+                this.max = Conversions.toByteBuffer(type, max);
```
if MetricsMode is != Full (eg Counts, None), no min/max should be stored.

again, this should be handled in defensive manner, instead of if-ing on a Full class

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcMetricsCollection.java:
```diff
@@ -129,7 +249,7 @@ public void testBasic()
         assertQuery("SELECT min(clerk) FROM tpch.tiny.orders", "VALUES '" + lowerBounds.get(7) + "'");
         assertQuery("SELECT min(shippriority) FROM tpch.tiny.orders", "VALUES " + lowerBounds.get(8));
-        assertQuery("SELECT min(comment) FROM tpch.tiny.orders", "VALUES '" + lowerBounds.get(9) + "'");
+        assertQuery("SELECT substring(min(comment), 1, 16) FROM tpch.tiny.orders", "VALUES '" + lowerBounds.get(9) + "'");
```
`lowerBounds.get(9)` is the actual, while `SELECT substring(min(comment), 1, 16) FROM tpch.tiny.orders` is the expected. The expected/actual are reversed here.

liqinrae replied:
@findepi I tried reversing the order for all the `assertQuery` in the test function but encountered failures when using `SELECT substring(min(comment), 1, 16) FROM tpch.tiny.orders` as the expected sql statement. Checked other test files and looks like they placed all `select` statement as the expected and result as the actual.

findepi replied:
we shouldn't use `assertQuery` at all, when we want to assert a value that we have at hand.

this is preexisting, so ok to followup

findepi reviewed (commented):
Please squash your commits.

When doing so, try not to change the base commit:

```
git rebase -i $(git merge-base head master)
```

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -261,10 +277,33 @@ private static int getIcebergId(OrcType orcColumn)
+            else if (metricsMode instanceof MetricsModes.Truncate) {
+                MetricsModes.Truncate truncateMode = (MetricsModes.Truncate) metricsMode;
+                int truncateLength = truncateMode.length();
+                switch (type.typeId()){
```
```suggestion
                switch (type.typeId()) {
```

cc @ksobolew , why not require whitespace before `{`?

ksobolew replied:
Checkstyle didn't complain? Oh, it didn't indeed... Looks like a gap in the configuration

ksobolew replied:
...or it's a bug in checkstyle. Whitespace before `{` is not allowed by our configuration, and checkstyle complains about it *except* after a `switch (...)`.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcMetricsCollection.java:
```diff
@@ -141,7 +261,8 @@ public void testBasic()
         assertQuery("SELECT max(shippriority) FROM tpch.tiny.orders", "VALUES " + upperBounds.get(8));
-        assertQuery("SELECT max(comment) FROM tpch.tiny.orders", "VALUES '" + upperBounds.get(9) + "'");
+        // the default truncate(16) mode would round up the last character of the maximum bound while keeping the first 15 characters identical
+        assertQuery("SELECT substring(max(comment), 1, 15) FROM tpch.tiny.orders", "VALUES '" + upperBounds.get(9).substring(0, 15) + "'");
```
assert on `upperBounds.get(9)` without taking substring

on the left side, use `substring(max(comment), 1, 15) || '...'` to construct the expected value

findepi commented:
@liqinrae have you had time to update this for my comments?

liqinrae commented:
> @liqinrae have you had time to update this for my comments?

@findepi Hey thanks for reviewing my changes. Updated the PR for your comment and also replaced `assertQuery` in the preexisting test class with the actual min/max value. Please review and let me know your thoughts. Thanks.
